### PR TITLE
feature (new Integration): OneSignal(cloud mode) destination Onboarding

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> Onboarding OneSignal(cloud mode) destination 

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/one_signal/transform.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+const identifyResponseBuilder = (message, { Config }) => {
+  const { appId, emailDeviceType, smsDeviceType } = Config;
+
+  if (!appId) {
+    throw new CustomError(
+      "[OneSignal]: appId is required for creating/adding a device",
+      400
+    );
+  }
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:67]: This is repeating.

File: v0/destinations/one_signal/transform.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+};
+const getExternalIdentifiersMapping = message => {
+  // Data Structure expected:
+  // context.externalId: [ {type: playerId, id: __id}]
+  const externalId = get(message, "context.externalId");
+  let playerId;
+  if (externalId && Array.isArray(externalId)) {
+    externalId.forEach(id => {
+      if (id.type === "playerId") {
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:51]:
```suggestion
      if (id.type === "onesignalPlayerId") {
```
We generally add destination name with the type.

File: v0/destinations/one_signal/transform.js [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+const populateDevicetype = (message, payload) => {
+  const integrationsObj = getIntegrationsObj(message, "one_signal");
+  const devicePayload = payload;
+  if (integrationsObj && integrationsObj.deviceType) {
+    devicePayload.device_type = integrationsObj.deviceType;
+    devicePayload.identifier = integrationsObj.identifier;
+  }
+};
+const getExternalIdentifiersMapping = message => {
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:44]: Do we need this function? We can directly call `getDestinationExternalID` function.

File: v0/destinations/one_signal/transform.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+  let { endpoint } = ENDPOINTS.IDENTIFY;
+
+  const playerId = getExternalIdentifiersMapping(message);
+  const tags = {};
+
+  const traits = getFieldValueFromMessage(message, "traits");
+  const traitsKey = Object.keys(traits);
+  traitsKey.forEach(key => {
+    tags[key] = traits[key];
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:77]: Put a check for strings only.

File: v0/destinations/one_signal/transform.js [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    payload.app_id = appId;
+    return responseBuilder(payload, endpoint);
+  }
+  populateDevicetype(message, payload);
+  const responseArray = [];
+  payload.tags = tags;
+  if (emailDeviceType) {
+    const emailDevicePayload = { ...payload };
+    emailDevicePayload.device_type = 11;
+    emailDevicePayload.identifier = getFieldValueFromMessage(message, "email");
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:98]: What if the email is not there?

File: v0/destinations/one_signal/transform.js [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    responseArray.push(responseBuilder(emailDevicePayload, endpoint));
+  }
+  if (smsDeviceType) {
+    const smsDevicePayload = { ...payload };
+    smsDevicePayload.device_type = 14;
+    smsDevicePayload.identifier = getFieldValueFromMessage(message, "phone");
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:104]: Same here.

File: v0/destinations/one_signal/transform.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    responseArray.push(responseBuilder(smsDevicePayload, endpoint));
+  }
+  if (payload.device_type) {
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:107]: We should check for `identifier` as well.

File: v0/destinations/one_signal/transform.js [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+      "[OneSignal]: userId is required for track events/updating a device",
+      400
+    );
+  }
+  endpoint = `${endpoint}/${appId}/users/${externalUserId}`;
+  const payload = {};
+  const tags = {};
+  if (event) {
+    tags[event] = true;
+  }
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:135]: We should throw an error if `event` is not defined.

File: v0/destinations/one_signal/transform.js [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+  const properties = get(message, "properties");
+  allowedProperties.forEach(propertyName => {
+    if (properties[propertyName] && eventAsTags && event) {
+      tags[`${event}_${[propertyName]}`] = properties[propertyName];
+    } else if (properties[propertyName]) {
+      tags[propertyName] = properties[propertyName];
+    }
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:142]:
```suggestion
    if (properties[propertyName]){
      if(eventAsTags){
            tags[`${event}_${[propertyName]}`] = properties[propertyName];
      }
      else
      {
            tags[propertyName] = properties[propertyName];
      }
    } 
```

File: v0/destinations/one_signal/config.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+const ENDPOINTS = {
+  IDENTIFY: {
+    endpoint: "/players"
+  },
+  TRACK: {
+    endpoint: "/apps"
+  },
+  GROUP: {
+    endpoint: "/players"
```
Comment (ItsSudip) [inline v0/destinations/one_signal/config.js:13]: For group calls, we can use the `apps` endpoint as we update the existing user with a tag.

File: v0/destinations/one_signal/transform.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    const smsDevicePayload = { ...payload };
+    smsDevicePayload.device_type = 14;
+    smsDevicePayload.identifier = getFieldValueFromMessage(message, "phone");
+    responseArray.push(responseBuilder(smsDevicePayload, endpoint));
+  }
+  if (payload.device_type) {
+    responseArray.push(responseBuilder(payload, endpoint));
+  }
+  return responseArray;
+};
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:205]: We should not create one user using group calls.

File: v0/destinations/one_signal/transform.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+    devicePayload.identifier = integrationsObj.identifier;
+  }
+};
+const getExternalIdentifiersMapping = message => {
+  // Data Structure expected:
+  // context.externalId: [ {type: playerId, id: __id}]
+  const externalId = get(message, "context.externalId");
+  let playerId;
+  if (externalId && Array.isArray(externalId)) {
+    externalId.forEach(id => {
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: removed

File: v0/destinations/one_signal/transform.js (hunk 2 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: removed

File: v0/destinations/one_signal/data/OneSignalGroupConfig.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    "required": false
+  },
+  {
+    "sourceKeys": "context.os",
+    "destKey": "device_os",
+    "sourceFromGenericMap": true,
+    "required": false
+  },
+  {
+    "sourceKeys": ["traits.address.country", "context.traits.address.country"],
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalGroupConfig.json]: Updated

File: v0/destinations/one_signal/data/OneSignalGroupConfig.json [hunk 14]
```diff
@@ -0,0 +1,38 @@
+[
+  {
+    "sourceKeys": "context.device.deviceModel",
+    "destKey": "device_model",
+    "sourceFromGenericMap": true,
+    "required": false
+  },
+  {
+    "sourceKeys": "context.os",
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalGroupConfig.json]: Corrected

File: v0/destinations/one_signal/data/OneSignalGroupConfig.json [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  {
+    "sourceKeys": "context.timezone",
+    "destKey": "timezone",
+    "sourceFromGenericMap": true,
+    "required": false
+  },
+  {
+    "sourceKeys": "groupId",
+    "destKey": "groupId",
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalGroupConfig.json]: done

File: v0/destinations/one_signal/data/OneSignalIdentifyConfig.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    "sourceFromGenericMap": true,
+    "required": false
+  },
+  {
+    "sourceKeys": "userId",
+    "destKey": "external_user_id",
+    "sourceFromGenericMap": true,
+    "required": false
+  }
+]
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalIdentifyConfig.json:37]: updated

File: v0/destinations/one_signal/transform.js (hunk 1 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/transform.js (hunk 3 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/transform.js (hunk 4 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Added

File: v0/destinations/one_signal/transform.js (hunk 5 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:93]: Added check

File: v0/destinations/one_signal/transform.js (hunk 6 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:107]: done

File: v0/destinations/one_signal/transform.js (hunk 7 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: done

File: v0/destinations/one_signal/transform.js (hunk 8 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:157]: done

File: v0/destinations/one_signal/transform.js (hunk 9 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: updated

File: v0/destinations/one_signal/config.js (hunk 10 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/config.js]: updated

File: v0/destinations/one_signal/transform.js (hunk 11 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:213]: updated

File: v0/destinations/one_signal/util.js [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * This function is used to populate device_type required for creating a device
+ * @param {*} message
+ * @param {*} payload
+ */
+const populateDeviceType = (message, payload) => {
+  const integrationsObj = getIntegrationsObj(message, "one_signal");
+  const devicePayload = payload;
+  if (integrationsObj && integrationsObj.deviceType) {
```
Comment (ItsSudip) [inline v0/destinations/one_signal/util.js:52]: Add check for integrationsObj.identifier as well.

File: v0/destinations/one_signal/util.js [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    if (message.channel === "mobile") {
+      devicePayload.device_type =
+        deviceTypeMapping[message.context?.device?.type?.toLowerCase()];
+      devicePayload.identifier = message.context?.device?.token
+        ? message.context?.device?.token
+        : message.context?.device?.id;
+    }
+    // Parsing the UA to get the browser info to map the device_type
+    if (message.channel === "web" && message.context?.userAgent) {
+      const browser = getBrowserInfo(message.context.userAgent);
```
Comment (ItsSudip) [inline v0/destinations/one_signal/util.js:74]: `getBrowserInfo` is not defined.

File: v0/destinations/one_signal/transform.js [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+      "Content-Type": "application/json"
+    };
+    if (eventType.toLowerCase() === "identify") {
+      response.method = defaultPostRequestConfig.requestMethod;
+    } else {
+      response.method = defaultPutRequestConfig.requestMethod;
+    }
+    response.body.JSON = removeUndefinedAndNullValues(payload);
+    return response;
+  }
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:40]: Throw an error if payload is not populated.

File: v0/destinations/one_signal/transform.js [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+ * It is used to edit the OneSignal tags using external_user_id.
+ * @param {*} message
+ * @param {*} param1
+ * @returns
+ */
+const trackResponseBuilder = (message, { Config }) => {
+  const { appId, eventAsTags, allowedProperties } = Config;
+  const event = get(message, "event");
+  let { endpoint } = ENDPOINTS.TRACK;
+  const externalUserId = getFieldValueFromMessage(message, "userId");
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:135]:
```suggestion
  const externalUserId = getFieldValueFromMessage(message, "userIdOnly");
```

File: v0/destinations/one_signal/transform.js [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+  const tags = {};
+  /* Populating event as true in tags.
+  eg. tags: {
+    "event_name": true
+  }
+  */
+  tags[event] = true;
+  // Populating tags using allowed properties(from dashboard)
+  const properties = get(message, "properties");
+  allowedProperties.forEach(item => {
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:159]: check if `allowedProperties` is defined and an array

File: v0/destinations/one_signal/transform.js [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+  const { appId, allowedProperties } = Config;
+  const groupId = getFieldValueFromMessage(message, "groupId");
+  if (!groupId) {
+    throw new CustomError(
+      "[OneSignal]: groupId is required for group events",
+      400
+    );
+  }
+  let { endpoint } = ENDPOINTS.GROUP;
+  const externalUserId = getFieldValueFromMessage(message, "userId");
```
Comment (ItsSudip) [inline v0/destinations/one_signal/transform.js:188]:
```suggestion
  const externalUserId = getFieldValueFromMessage(message, "userIdOnly");
```

File: v0/destinations/one_signal/data/OneSignalIdentifyConfig.json [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    "destKey": "country",
+    "required": false
+  },
+  {
+    "sourceKeys": "context.timezone",
+    "destKey": "timezone",
+    "required": false
+  },
+  {
+    "sourceKeys": "userId",
```
Comment (ItsSudip) [inline v0/destinations/one_signal/data/OneSignalIdentifyConfig.json:41]:
```suggestion
    "sourceKeys": "userIdOnly",
```
As we are already setting the anonymousId as a tag

File: v0/destinations/one_signal/transform.js (hunk 21 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:159]: Do we need to add this check if it is taken from dashboard only.

File: v0/destinations/one_signal/transform.js [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+  // Mapping app_id
+  payload.app_id = appId;
+
+  // update the timestamp to unix timeStamp
+  if (payload.created_at) {
+    payload.created_at = toUnixTimestamp(payload.created_at);
+  }
+  if (payload.last_active) {
+    payload.last_active = toUnixTimestamp(payload.last_active);
+  }
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Done

File: v0/destinations/one_signal/util.js (hunk 17 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/util.js]: updated

File: v0/destinations/one_signal/util.js (hunk 18 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/util.js:81]: Done

File: v0/destinations/one_signal/transform.js (hunk 19 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:39]: Done

File: v0/destinations/one_signal/transform.js (hunk 20 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/transform.js (hunk 22 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/data/OneSignalIdentifyConfig.json (hunk 23 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalIdentifyConfig.json]: Updated

File: v0/destinations/one_signal/transform.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Done

File: v0/destinations/one_signal/util.js (hunk 17 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/util.js]: Done

File: v0/destinations/one_signal/util.js (hunk 18 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/util.js:81]: Updated

File: v0/destinations/one_signal/transform.js (hunk 19 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js:39]: Done

File: v0/destinations/one_signal/transform.js (hunk 20 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/transform.js (hunk 22 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Updated

File: v0/destinations/one_signal/data/OneSignalIdentifyConfig.json (hunk 23 above)
Comment (ujjwal-ab) [reply v0/destinations/one_signal/data/OneSignalIdentifyConfig.json]: Updated

File: v0/destinations/one_signal/transform.js (hunk 21 above)
Comment (ItsSudip) [reply v0/destinations/one_signal/transform.js]: As it is coming from some other services we can't rely on them. We should check if the correct value is coming or not.

Comment (ujjwal-ab) [reply v0/destinations/one_signal/transform.js]: Done

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**